tree/ntuple/v7/inc/ROOT/RCluster.hxx

```diff
@@ -0,0 +1,188 @@
+/// \file ROOT/RCluster.hxx
+/// \ingroup NTuple ROOT7
+/// \author Jakob Blomer <jblomer@cern.ch>
+/// \date 2020-03-11
+/// \warning This is part of the ROOT 7 prototype! It will change without notice. It might trigger earthquakes. Feedback
+/// is welcome!
+
+/*************************************************************************
+ * Copyright (C) 1995-2020, Rene Brun and Fons Rademakers.               *
+ * All rights reserved.                                                  *
+ *                                                                       *
+ * For the licensing terms see $ROOTSYS/LICENSE.                         *
+ * For the list of contributors see $ROOTSYS/README/CREDITS.             *
+ *************************************************************************/
+
+#include <ROOT/RNTupleUtil.hxx>
+
+#include <unordered_map>
+#include <unordered_set>
+#include <vector>
+
+#ifndef ROOT7_RCluster
+#define ROOT7_RCluster
+
+namespace ROOT {
+namespace Experimental {
+namespace Detail {
+
+
+// clang-format off
+/**
+\class ROOT::Experimental::ROnDiskPage
+\ingroup NTuple
+\brief A page as being stored on disk, that is packed and compressed
+
+Used by the cluster pool to cache pages from the physical storage. Such pages generally need to be
+uncompressed and unpacked before they can be used by RNTuple upper layers.
+*/
+// clang-format on
+class ROnDiskPage {
+private:
+   /// The memory location of the bytes
+   const void *fAddress = nullptr;
+   /// The compressed and packed size of the page
+   std::size_t fSize = 0;
+
+public:
+   /// On disk pages within a page source are identified by the column and page number. The key is used for
+   /// associative collections of on disk pages.
+   struct Key {
+      DescriptorId_t fColumnId;
+      NTupleSize_t fPageNo;
```

Comment:
Perhaps we need a better name.  The `NTupleSize_t` is supposed to be an integer large enough to store a _count_ of any sort in an RNTuple. For instance, the number of elements in a column, the number of fields in an RNTuple, the number of clusters, the number of pages of a column, ...

In this case, a `size_t` would work in practice but semantically it would be odd because it would make the type machine dependent.

Reply:
Changed to `std::uint64_t`